### ISSUE TITLE
ESM/CJS Bundling results in double import

### ISSUE THREAD

**Issue by timbrinded** [polkadot-js/api#5636]
* **I'm submitting a ...**

  - [X] Bug report
  - [ ] Feature request
  - [ ] Support request
  - [ ] Other

* **What is the current behavior and expected behavior?**

When importing Polkadot{.js} packages, it results in both ESM and CJS getting imported into the project. This produces a wall of warning on applications that are dependent on polkadot.js, but maybe will also cause unexpected effects down the line?

```
@polkadot/util has multiple versions, ensure that there is only one installed.
Either remove and explicitly install matching versions or dedupe using your package manager.
The following conflicting packages were found:
	esm 12.1.2	node_modules/@polkadot/util/
	cjs 12.1.2	node_modules/@polkadot/util/cjs
@polkadot/api has multiple versions, ensure that there is only one installed.
Either remove and explicitly install matching versions or dedupe using your package manager.
The following conflicting packages were found:
	esm 10.5.1	node_modules/@polkadot/api/
	cjs 10.5.1	node_modules/@polkadot/api/cjs
@polkadot/rpc-augment has multiple versions, ensure that there is only one installed.
Either remove and explicitly install matching versions or dedupe using your package manager.
The following conflicting packages were found:
	esm 10.5.1	node_modules/@polkadot/rpc-augment/
	cjs 10.5.1	node_modules/@polkadot/rpc-augment/cjs
@polkadot/keyring has multiple versions, ensure that there is only one installed.
Either remove and explicitly install matching versions or dedupe using your package manager.
The following conflicting packages were found:
	esm 12.1.2	node_modules/@polkadot/keyring/
	cjs 12.1.2	node_modules/@polkadot/keyring/cjs
```

Whilst dual building repos is beyond tricky, I have seen some [other projects](https://github.com/vitest-dev/vitest) manage this area with help of middleware. `tsup` or `rollup` help in this area, with the former providing a pretty low-touch solution. I think the issue here is that both ESM and CJS share the `.js` extension. 

What i've done in my dual-publish repo is to have them published alongside each other, using the `.cjs` extension to denote the old format. Coupled with the following in my `package.json`, it means the correct entry point is selected:

```
  "exports": {
    ".": {
      "require": "./dist/index.cjs",
      "import": "./dist/index.js",
      "types": "./dist/index.d.ts"
    }
  },
  "module": "./dist/index.js",
  "main": "./dist/index.cjs",
  "types": "./dist/index.d.ts",
```

* **What is the motivation for changing the behavior?**

Reduce the conflicts in my team's repos which use polkadot.js

* **Please tell us about your environment:**

  - Version: `pnpm: 8.4.0`
  - Environment:

    - [X] Node.js
    - [ ] Browser
    - [ ] Other (limited support for other environments)

  - Language:

    - [ ] JavaScript
    - [X] TypeScript (include tsc --version)
    - [ ] Other

**Comment by jacogr** [comment]
Import maps are being used. Even with `.cjs` & `.mjs` extensions, the same warning will be shown - there are different parts defined, e.g. see `rpc-augment` above -

```json
"exports": {
    "./cjs/package.json": "./cjs/package.json",
    "./cjs/*": "./cjs/*.js",
    ".": {
      "types": "./index.d.ts",
      "require": "./cjs/index.js",
      "default": "./index.js"
    },
```

The warning there is explicit to warn against different paths being used inside the same code (or dependencies).

**Comment by timbrinded** [comment]
Ah thanks, didnt know that about `types` being first!

So it sounds like this can't be fixed then unless node changes then? if that's the case i'll close the issue

**Comment by jacogr** [comment]
It is a massive PITA. Since this is a classic case of the [dual package hazard](https://nodejs.org/api/packages.html#dual-package-hazard), the options really are -

1. Publish all `@polkadot` packages only as CJS (or a stub that still always points to CJS even when ESM is used).
2. Publish `@polkadot` only as ESM (this basically forces everybody to package the same and rewrite all cjs scripts out there)

(1 & 2 above are really the only _real_ solution - there is only a single version of the code present)
 
3. [Remove the warning](https://github.com/polkadot-js/common/blob/b73b96e412cd1333143efe2dc64f1269c12756e8/packages/util/src/versionDetect.ts#L120) about dual usage (it will still occur, but we are basically just "hiding" it)
4. Get all `@polkadot` users to also publish ESM/CJS (There are a variety out there, [some only CJS](https://github.com/AcalaNetwork/acala.js), [some only ESM](https://github.com/logion-network/logion-api), [some ESM and CJS](https://github.com/fragcolor-xyz/fragnova/tree/devel/js/api-augment)).

At this point I'm at "5. Do nothing and hope people move to ESM".

There is also a "6. Alias everything in your bundler" option, which just pushes the effort to _every_ single user. 

(Where I actually use _all_ the various [types packages](https://github.com/polkadot-js/apps/blob/155a23e97823aeb88b5dcb455ea2afa80f8b9e85/packages/apps-config/package.json#L18-L56) I use a pre-build step to flatten all, but I wouldn't call that a scalable-solution - it just means consumers of that specific config never runs into this - it is similar to point 6 above)

TL;DR

**Comment by jacogr** [comment]
7. Add a `module` key to the exports map (in addition to `import` & `require`), at least for (some) bundlers this would force a single version (It is the same as 6 above, just here with native `webpack` & `rollup` support via maps). See [https://webpack.js.org/guides/package-exports/]

I have really never seen this key used in the wild in any export map, but could make life simpler for those using bundlers. (According to the documentation - stumble on it from a webpack issue that asked for a key for this exact dual-package-hazard reason)

**Comment by jacogr** [comment]
Going to close this one as-is. 

Apart from adding another key, aka `module` added in the linked PR to all export maps, used by at least `rollup` & `webpack` bundlers (it will be available in the next version bumps across the line), there is not much that reasonably can be done with dual packages. The addition should improve things for at least some users out there.

Certainly a "hazard".

**Comment by timbrinded** [comment]
The detail is much appreciated and most informative!
